File: python/89_Gray_Code.py
#!/usr/bin/python


# 2 successive values differ in 1 bit
class Solution:
    def grayCode(self, n):
        if n < 0:
        	return []
        res = [0]
        shift = 1
        for i in range(n):
        	for j in range(len(res) - 1, -1, -1):
        		res.append(res[j] + shift)
        	shift = shift << 1
        return res

def main():
    s = Solution()
    print s.grayCode(3)

if __name__ == "__main__":
    main()

File: c++/89_Gray_Code.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> grayCode(int n) {
    	vector<int> res;
    	if (n < 0)
    		return res;
    	res.push_back(0);
    	int shift = 1;
    	for (int i = 0; i < n; i++) {
    		for (int j = res.size() - 1; j >= 0; j--)
    			res.push_back(res[j] + shift);
    		shift <<= 1;
    	}
    	return res;
    }
};